sdk/python/feast/infra/online_stores/contrib/bigtable_online_store/bigtable.py

```diff
@@ -0,0 +1,301 @@
+from pydantic.typing import Literal
+
+from feast import Entity, FeatureView, feature_view, utils
+from feast.infra.online_stores.helpers import compute_entity_id
+from feast.infra.online_stores.online_store import OnlineStore
+from feast.protos.feast.types.EntityKey_pb2 import EntityKey as EntityKeyProto
+from feast.protos.feast.types.Value_pb2 import Value as ValueProto
+from feast.protos.feast.types.Value_pb2 import ValueType
+from feast.repo_config import FeastConfigBaseModel, RepoConfig
+from feast.usage import get_user_agent, log_exceptions_and_usage, tracing_span
+
+logger = logging.getLogger(__name__)
+
+# Number of mutations per BigTable write operation we're aiming for. The official max is
+# 100K; we're being conservative.
+MUTATIONS_PER_OP = 50_000
+# The Bigtable client library limits the connection pool size to 10. This imposes a
+# limitation to the concurrency we can get using a thread pool in each worker.
+BIGTABLE_CLIENT_CONNECTION_POOL_SIZE = 10
+
+
+class BigTableOnlineStoreConfig(FeastConfigBaseModel):
+    """Online store config for GCP BigTable"""
+
+    type: Literal["bigtable"] = "bigtable"
+    """Online store typee selector"""
+
+    project_id: Optional[StrictStr] = None
+    """(optional) GCP Project ID"""
+
+    instance: StrictStr
+    """The BigTable instance's ID"""
+
+    max_versions: int = 2
+    """The number of historical versions of data that will be kept around."""
+
+
+class BigTableOnlineStore(OnlineStore):
+    _client: Optional[bigtable.Client] = None
+
+    feature_column_family: str = "features"
+
+    @log_exceptions_and_usage(online_store="bigtable")
+    def online_read(
+        self,
+        config: RepoConfig,
+        table: FeatureView,
+        entity_keys: List[EntityKeyProto],
+        requested_features: Optional[List[str]] = None,
+    ) -> List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]]:
+        feature_view = table
+        bt_table_name = self._get_table_name(config=config, feature_view=feature_view)
+
+        client = self._get_client(online_config=config.online_store)
+        bt_instance = client.instance(instance_id=config.online_store.instance)
+        bt_table = bt_instance.table(bt_table_name)
+        row_keys = [
+            self._compute_row_key(
+                entity_key=entity_key,
+                feature_view_name=feature_view.name,
+                config=config,
+            )
+            for entity_key in entity_keys
+        ]
+
+        batch_result: List[
+            Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]
+        ] = []
+
+        # TODO: read all the rows in a single call instead of reading them sequentially
+        for row_key in row_keys:
+            res = {}
+            # TODO: use filters to reduce the amount of data transfered and skip
+            # unnecessary columns.
+            row = bt_table.read_row(row_key)
+
+            if row is None:
+                batch_result.append((None, None))
```

Comment:
Note to self: if row is None, then the rest of the code below needs to be bypassed using `continue`.